TBRANDS-106 & TBRANDS-126 : Product Gallery Desktop View

ecarlisle opened the pull request:
## Description
When the browser viewport is desktop-sized, clicking on images within the Product Gallery will produce a "Focused View." This view is essentially a modal overlay, allowing the user to browse larger product images. 
* [Figma Prototype](https://www.figma.com/proto/vWVGgZfZVBdSX7ACo3kmjr/Enhanced-Styling-Design-System?node-id=8294%3A130007&scaling=scale-down&page-id=8294%3A119303&starting-point-node-id=8294%3A119408&show-proto-sidebar=1)
* [Figma Design](https://www.figma.com/file/vWVGgZfZVBdSX7ACo3kmjr/Enhanced-Styling-Design-System?node-id=8020%3A112533)

## Jira Ticket

- Original Ticket : [TBRANDS-106](https://arcpublishing.atlassian.net/browse/TBRANDS-106)
- Follow-up Ticket: [TBRANDS-126](https://arcpublishing.atlassian.net/browse/TBRANDS-126)


## Acceptance Criteria

### From TBRANDS-106
1. When a consumer user clicks on any image rendered as part of the Product Gallery, the Focused View deploys.
2. The Focused View fills the user's browser’s viewport. (Not Fullscreen API)
3. A user can scroll vertically through all gallery images.
4. A user can navigate to a specific image by clicking on a thumbnail in the left rail. 
5. A user sees a thumbnail highlighted that corresponds to the image they are viewing.
6. A user can close the Focused View by clicking on the close icon found at the top right.

### From Figma Design:

#### BEHAVIOR (Responsive)
On desktop, clicking on a gallery media item will open the Gallery Focus for a more detailed view.

#### IMAGE
1. Aspect ratio should be square (1:1), at least 2000px x 2000px
2. Image should fill 100% width of the dedicated content area above, height extending beyond the viewport height
3. Scrolling over focused content area will reveal additional media

#### SCROLL (over focused media) 
* All media in the Gallery Focus View can be accessed via vertical scroll. 
* Scrolling over the focused media will reveal the next media item in the carousel. 
* Once that next item takes up the majority screen height, the corresponding thumbnail will be selected.
* If the thumbnail is not present in current carousel view, the carousel will shift to ensure the selected thumbnail is in view.
#### Thumbnails
1. Aspect ratio should be square (1:1), 100x100
2. Height of Thumbnail Bar should not exceed viewport height - 80px. Number of thumbnails displayed at a time is dependent on the viewport height. An additional thumbnail can be shown as long as the gap between elements and padding above and below is maintained.
3. Thumbnail bar is fixed so should maintain it’s position when focused area is vertically scrolled.

### Assumed Out of Scope
1. Video integration
2. Image zoom control
3. Product color selection

## Test Steps

1. `git checkout TBRANDS-106-product-gallery-focus-view` on local `arc-themes-blocks` repository.
2. `git checkout TBRANDS-106-product-gallery-focus-view` on local `arc-themes-feature-pack` repository.
3. In local `arc-themes-feature-pack` repository, enable the following lines to their local repositories:
    a. `THEME_COMPONENTS_REPO` (needed for change in `Image` component).
    b. `THEMES_BLOCKS_REPO`
4. Start feature pack locally with `npx fusion start -f -l @wpmedia/product-gallery-block`.
5. Create a new page in PageBuilder Editor:
    a. Set global content of page to `commerce-product` using `bounce-running-shoe` as the SKU.
    b. Add "Product Gallery - Arc Block" to page.
    c. Set "Display the product's featured image" to true.
6. Publish the page and view locally for best debugging.

### Example Screens

#### Product Gallery at fullscreen (without Focused View overlay).
<img width="1417" alt="image" src="https://user-images.githubusercontent.com/452134/193931467-8cb4d280-651b-4e57-a9a5-67d51ac99ca5.png">

#### Focused View Overlay (Number of thumbnails statically set to 3)
<img width="1623" alt="image" src="https://user-images.githubusercontent.com/452134/193934690-6d20b559-b383-4f9a-8a84-1ab60118522d.png">

#### Thumbnail Bar States (3)
![image](https://user-images.githubusercontent.com/452134/193937268-ed4978db-6dc2-4fa4-ab91-a164b2ff30b2.png)


## Dependencies or Side Effects

Feature Pack Styles - https://github.com/WPMedia/arc-themes-feature-pack/pull/359

## Author Checklist

_The author of the PR should fill out the following sections to ensure this PR is ready for review._

- [ ] Confirmed all the test steps a reviewer will follow above are working.
- [ ] Confirmed there are no linter errors. Please run `npm run lint` to check for errors. Often, `npm run lint:fix` will fix those errors and warnings.
- [ ] Ran this code locally and checked that there are not any unintended side effects. For example, that a CSS selector is scoped only to a particular block.
- [ ] Confirmed this PR has reasonable code coverage. You can run `npm run test:coverage` to see your progress.
  - [ ] Confirmed this PR has unit test files
  - [ ] Ran `npm run test`, made sure all tests are passing
  - [ ] If the amount of work to write unit tests for this change are excessive,
        please explain why (so that we can fix it whenever it gets refactored).
- [ ] Confirmed relevant documentation has been updated/added.

## Reviewer Checklist

_The reviewer of the PR should copy-paste this template into the review comments on review._

- [ ] Linting code actions have passed.
- [ ] Ran the code locally based on the test instructions.
  - [ ] I don’t think this is needed to be tested locally. For example, a padding style change (storybook?) or a logic change (write a test).
- [ ] I am a member of the engine theme team so that I can approve and merge this. If you're not on the team, you won't have access to approve and merge this pr.
- [ ] Looked to see that the new or changed code has code coverage, specifically. We want the global code coverage to keep on going up with targeted testing.

ecarlisle committed:
Update gallery block styles with focus view hooks.

ecarlisle committed:
Integrate focus view into gallery.

ecarlisle committed:
Add accessibility labels to ProductFocvusView buttons.

ecarlisle committed:
Add test for Product Focus View.

ecarlisle committed:
Change mock data for Product Focus view tests.

ecarlisle committed:
Lint corrections in Product focus view tests.

ecarlisle committed:
Add new styles for Storybook.

ecarlisle committed:
Set thumbnail bar image limit to 3.

ecarlisle committed:
Revert "Set thumbnail bar image limit to 3."

This reverts commit 4b93c817bd7b0f3aabd996969cb8ee2b252b78df.

ecarlisle committed:
Set smaller thumbnailBarItemLimit on ProductFocusView for local testing.

ecarlisle committed:
Change test mock data to match change in thumbnailBarItemLimit.

ecarlisle committed:
Change test to match new number of mock data test images.

ecarlisle committed:
Fix Lint errors in ProductListView Sass.

ecarlisle committed:
Merge branch 'arc-themes-release-version-2.0.3' into TBRANDS-106-product-gallery-focus-view

ecarlisle committed:
Account for change in Arc Commerce Schema for image location.

nschubach commented on .storybook/themes/commerce.scss:
```diff
@@ -803,6 +803,75 @@
+						paragraph: (
+							font-family: var(--font-family-primary),
+							font-size: var(--heading-level-6-font-size),
+							font-weight: var(--global-font-size-4),
```
Font weights should use font-weight values.

nschubach commented on blocks/product-gallery-block/features/product-gallery/_children/ProductFocusView/index.jsx:
```diff
@@ -0,0 +1,138 @@
+					{productImages.map((image, index) => (
+						<div
+							style={{ flexBasis: "100vh", flexShrink: 0 }}
+							key={image._id}
```
`key`s should be unique:
![image](https://user-images.githubusercontent.com/2287238/194131200-723f4b54-7247-498c-9cae-ae42ce1f6c89.png)

Should just be a matter of appending the `_id`: `focus_main_view_${image._id}`

ecarlisle replied:
@nschubach : Good call! Made this change.

nschubach commented on blocks/product-gallery-block/features/product-gallery/_children/ProductFocusView/index.jsx:
```diff
@@ -0,0 +1,138 @@
+										image._id === productImages[selectedImageIndex]._id ? "--selected" : ""
+									}`}
+									height={100}
+									key={image._id}
```
`key`s should be unique.

Should just be a matter of appending the `_id`: `focus_view_thumbnail_${image._id}`

ecarlisle replied:
@nschubach : Good call! Made this change.
I think the image `HY6LDPEW4BBFDLBYD4S3S7LZ3E` is coming from the original Product Gallery tests and that ID is being used multiple times, but I'm sure there's a way to make it unique.

nschubach commented on blocks/product-gallery-block/features/product-gallery/_children/ProductFocusView/index.jsx:
```diff
@@ -0,0 +1,138 @@
+							<Icon name="ChevronUp" />
+						</Button>
+					) : (
+						<div className={`${BLOCK_CLASS_NAME}__focus-view-thumbnail-bar-spacer`} />
```
Instead of a spacer, why don't we make this container(`focus-view-thumbnail-bar`) a grid and set the row height?  The "padding" could be row 1, selected text could be row 2, button in row 3, stack in row 4, button in row 5, and the bottom padding is row 6.

```
product-gallery-focus-view-thumbnail-bar: (
    display: grid,
    grid-template-rows: var(--global-spacing-7) auto 1fr auto 1fr var(--global-spacing-7),
    grid-template-columns: var(--global-spacing-6) auto var(--global-spacing-6),
)
```

nschubach commented on .storybook/themes/commerce.scss:
```diff
@@ -803,6 +803,75 @@
+					padding-right: var(--global-spacing-6),
+					padding-bottom: var(--global-spacing-7),
+					padding-left: var(--global-spacing-6),
+					text-align: center,
```
Technically this (`text-align`) should be on the paragraph.  If you want to center align the elements, with flex/grid, we should be using `align-items`/`justify-content` notation.

nschubach commented on .storybook/themes/commerce.scss:
```diff
@@ -803,6 +803,75 @@
+					border-width: 0.5px,
+					opacity: 0.75,
+				),
+				product-gallery-focus-view-thumbnail-image-selected: (
+					border-color: var(--global-black),
+					border-style: var(--global-border-style-1),
+					border-width: 0.5px,
+				),
```
If selected is a modified state, you should only have to set the `border-color`

ecarlisle committed:
Account for change in Arc Commerce Schema for respective tests.

nschubach commented on blocks/product-gallery-block/features/product-gallery/default.jsx:
```diff
@@ -43,63 +47,75 @@ export function ProductGalleryDisplay({
-				);
-			})}
-		</Carousel>
+		<div>
```
Why was this added?

nschubach commented on blocks/product-gallery-block/features/product-gallery/default.jsx:
```diff
@@ -43,63 +47,75 @@ export function ProductGalleryDisplay({
-				);
-			})}
-		</Carousel>
+		<div>
```
This can be a fragment to reduce DOM elements.

ecarlisle replied:
Good call! Changed to a fragment.

ecarlisle committed:
Fix merge conflict.

nschubach commented on blocks/product-gallery-block/features/product-gallery/default.jsx:
```diff
@@ -137,8 +153,6 @@ function ProductGallery({ customFields }) {
 	);
 }
 
-ProductGallery.label = "Product Gallery – Arc Block";
-
```
Did you mean to remove this?

ecarlisle replied:
Thanks! Unintended. I have replaced it.

ecarlisle committed:
Ensure keys are unique in the ProductFocusView.

ecarlisle committed:
Add label text back for Product Gallery.

ecarlisle committed:
Use fragment instead of <div> to surround Carousel and ProductFocusView.

ecarlisle reviewed (commented):
Refreshing comments if I started a review rather than adding comments.

ecarlisle commented:
> I know this is WIP, but are you adding a storybook item for this?

@nschubach : Yes. I've listed that under the "Pending, Incomplete Items" section above. I increased the title size and placed it just before the Acceptance Criteria to give it more attention.

JackHowa commented:
@ecarlisle 

I see the incomplete items. What have you seen not working here? what's expected? I'm also curious what you've tried so I can help you 

> 
> Pending, Incomplete Items
> The Product Focus View is not yet "Desktop Only." Click events on images need to render the overlay only when the viewport is desktop width or higher.
> Sass tokens have been set on default. Some/all may need to be moved to desktop depending on how the "Desktop Only" behavior is implemented.
> Storybook implementation of the Desktop view appears to be incomplete, though tested styles in arc-themes-feature-pack have also been placed in the Storybook commerce styles.
> The state of the thumbnail bar and selected thumbnail needs to change when a different product image scrolls is scrolled into view and "takes up the majority screen height." This is probably best accomplished using the Intersection Observer API.
> The number of Thumbnail images displayed depends on the height of the viewport. Currently that number is static and needs to be dynamic based on viewport height. is is probably also best accomplished using the Intersection Observer API.

ecarlisle committed:
Segment ProductFocusView into separate child component files.

ecarlisle committed:
Update CSS

ecarlisle committed:
MOved mock data for tests.

ecarlisle committed:
Component updates.

ecarlisle committed:
Add/update test files.

ecarlisle committed:
Set desktop threshold condition for Product Focus View.

ecarlisle committed:
Merge branch 'arc-themes-release-version-2.0.3' into TBRANDS-106-product-gallery-focus-view

ecarlisle committed:
Take windowResize function out of useEffect.

ecarlisle committed:
Update to tests

matthewroach committed:
Increase test coverage

matthewroach committed:
Merge branch 'arc-themes-release-version-2.0.3' into TBRANDS-106-product-gallery-focus-view

Commit:
Merge branch 'arc-themes-release-version-2.0.3' into TBRANDS-106-product-gallery-focus-view

matthewroach committed:
PR Updates

Add translations for next and previous thumbnails buttons
Fix styling issues, and remove inline styles

matthewroach committed:
Remove additional comments

matthewroach committed:
Update close button styles and add translation for label

matthewroach committed:
Merge branch 'arc-themes-release-version-2.0.3' into TBRANDS-106-product-gallery-focus-view

nschubach commented:
![image](https://user-images.githubusercontent.com/2287238/198738802-467fc32f-c607-432e-8156-3a8aa52c2198.png)
It seems that the collapse arrows on the Product-Content block are bleeding through... are we considering fixing that with this?

nschubach commented:
Close button appears to have moved below the content?
![image](https://user-images.githubusercontent.com/2287238/198739462-cb597f0d-ab3d-41bc-a26f-9b9e8375c8d6.png)

matthewroach commented:
@nschubach - Can you pull both the blocks branches and feature pack branches again - the issues you are seeing were addressed.

Feature Pack PR is - https://github.com/WPMedia/arc-themes-feature-pack/pull/359

nschubach commented on blocks/product-gallery-block/features/product-gallery/_children/ThumbnailBar/index.jsx:
```diff
@@ -0,0 +1,146 @@
+		thumbnailBarItemLimit < images.length && thumbnailBarStartIndex !== 0;
+
+	const windowResize = () => {
+		// Math
```
```suggestion
```

nschubach committed:
Update blocks/product-gallery-block/features/product-gallery/_children/ThumbnailBar/index.jsx

nschubach reviewed (approved):
Not sure I like the hardcoded desktop width (I think it could be just a styled `display:none` on this component) but this will do.
